### PR TITLE
Blog post: Programming beyond the comfort zone (1)

### PR DESCRIPTION
**Pull request by roseritter**
@erikmueller I already corrected a few spelling mistakes and will now comment a few more things for you to revise;)

### CODE REVIEW

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+Besides the usual suspects like `express` and `hapi`, there are a lot of things that are fun to play with like zeit's `micro`.
+Although having some different concepts, they all have something in common: They're (of course) JavaScript frameworks.
+Now, if you want to learn more about the HTTP stack or the request and response cycle, you should definitely check out a lot of different frameworks since different concepts give you different insights and perspectives of the inner workings.
+
+So why should you start programming in another language?
+
+Let's take a step back from the frameworks and look at the language itself.
+JavaScript claims to be both, object-oriented (with prototypal inheritance) and functional.
+And since it's both, it's neither nor to a satisfying extend.
+Other languages like Elixir (which we're going to talk about in a bit) make it much easier to use functional language features and methods such as *destructuring*, *pattern matching* or *recursion*.
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: add link: http://elixir-lang.org/

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+### Enter: Elixir and Phoenix
+
+This is rather a teaser that should make you curious about trying out something new and leaving your comfort zone, than a full blown tutorial or language reference.
+This first part will take a look at how **Elixir** might help you to rethink programming patterns leading to cleaner code and continue with what this **Phoenix** thing actually is.
+
+### The Platform
+
+Before we explore some Elixir code, let's start with a naive JS example of populating an array - mutation style.
+Don't fraud, I've seen a lot of people (including myself) starting like this (which is totally fine).
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: fraud or frown?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+my_numbers = [-2, 3, 5, -34, 0, 32]
+@@ -48,16 +48,16 @@ result = my_numbers
+
+```
+
+This pretty much resembles the JavaScript solution, right?!
+Almost!
+Since Elixir is based on immutable data, you could not have started like in the first JavaScript solution.
+You cannot simply mutate a data structure in Elixir!
+What you would have done, is to create a new list with every value from the old one that you see fits.
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: I don't quite understand this sentence

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+Besides that, there are some other oddities.
+The `|>` is called a pipe operator and passes the value as the first argument to the next function.
+While we're using the `fn` keyword to define an anonymous function as the `filter` expression and explicitly naming the variable `num`, we're shortening this for the `map` function and just say: "*This function uses the first argument (`&1`) it receives and doubles it*".
+
+Let's take the example from above again.
+In contrast to JavaScript, we can use some sugar in Elixir called a `for` comprehension in conjunction with a filter to loop over the enumerable and create a new one (with all values > 0 doubled).
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: maybe a little explanation that sugar is a framework?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+my_numbers = [-2, 3, 5, -34, 0, 32]
+result = for n <- my_numbers, n > 0, do: n * 2
+```
+
+Note that the last statement of every Elixir function is automatically returned creating a new list bound to `result`.
+What actually looks like an assignment `=` is actually a `match`.
+Trying to match `result = 4` afterwards will give you an error, since the compiler tries to match `[6, 10, 64] = 4`.
+This might look strange at first, yet makes more sense once you grasp concepts like pattern matching or destructuring.
+
+Since we now have seen some language features, I have to bore you with some facts.
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: Maybe better:
After looking at the above described language features, I cannot circumvent a few basic (and maybe boring {% emoji wink %}) facts.

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+Elixir runs on the Erlang VM (BEAM) and gives you the great ecosystem companies like *Heroku* and *WhatsApp* use to build their products.
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: maybe add a link to Erlang VM (BEAM)?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+In the JavaScript world, there are a gazillion web frameworks available and every week there are more coming.
+Besides the usual suspects like `express` and `hapi`, there are a lot of things that are fun to play with like zeit's `micro`.
+Although having some different concepts, they all have something in common: They're (of course) JavaScript frameworks.
+Now, if you want to learn more about the HTTP stack or the request and response cycle, you should definitely check out a lot of different frameworks since different concepts give you different insights and perspectives of the inner workings.
+
+So why should you start programming in another language?
+
+Let's take a step back from the frameworks and look at the language itself.
+JavaScript claims to be both, object-oriented (with prototypal inheritance) and functional.
+And since it's both, it's neither nor to a satisfying extend.
```
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: @erikmueller I've never seen this constellation in English: "neither nor"
Should be something like "neither this nor that". But what would it be?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+You might shift from doing everything with a `forEach` loop to solving pretty much every problem with `reduce`.
+This paradigm shift not only broadens your programming knowledge, but lets you view problems (and more importantly their solutions) from a different angle.
+On top of that, it improves code readability - at least most of the time {% emoji wink %}.
+
+Now, what if I told you that you can learn all of this by using a web framework (and language) that is not only fun to work with but also faster and less error prone than all of the before mentioned?
+
+### Enter: Elixir and Phoenix
+
+This is rather a teaser that should make you curious about trying out something new and leaving your comfort zone, than a full blown tutorial or language reference.
+This first part will take a look at how **Elixir** might help you to rethink programming patterns leading to cleaner code and continue with what this **Phoenix** thing actually is.
```
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: add link to: http://www.phoenixframework.org/ ?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+```elixir
+my_numbers = [-2, 3, 5, -34, 0, 32]
+@@ -48,16 +48,16 @@ result = my_numbers
+
+```
+
+This pretty much resembles the JavaScript solution, right?!
+Almost!
+Since Elixir is based on immutable data, you could not have started like in the first JavaScript solution.
```
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: @erikmueller missing word behind "like" ? >> maybe like that?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 3 above)
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: What you would have done is to create a new list with every value from the old one that you see fit.

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+Note that the last statement of every Elixir function is automatically returned creating a new list bound to `result`.
+What actually looks like an assignment `=` is actually a `match`.
+Trying to match `result = 4` afterwards will give you an error, since the compiler tries to match `[6, 10, 64] = 4`.
+This might look strange at first, yet makes more sense once you grasp concepts like pattern matching or destructuring.
+
+Since we now have seen some language features, I have to bore you with some facts.
+Elixir runs on the Erlang VM (BEAM) and gives you the great ecosystem companies like *Heroku* and *WhatsApp* use to build their products.
+It kind of relates to Erlang like Scala relates to Java (JVM).
+Likewise, you can import Erlang modules and use them right away.
+Elixir's syntax is concise and reminds of the Ruby programming language due to the fact that the creator [José Valim](https://github.com/josevalim) is a very active member of the ruby community.
```
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: @erikmueller could you spend one teaser phrase to the follow-up post?

Other than that: I LOVE IT! Very nicely written!

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 8 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: true, just did these links in the second post. Forgot them here

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 1 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: true, just did these links in the second post. Forgot them here

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 2 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: frown ^^'

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 9 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: don't know if 'that' is needed but if something is missing it would be 'that'

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 3 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: You would have created a new list containing every value from the old one that you deem worthy. xD?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 4 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: sugar is what is written there: sugar. It's no framework. It's basically a more elegant alias to things that could also be done in an unelegant way

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 5 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: then without the thing in parathesis

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 6 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: We can add a link to `erlang` there's no link to BEAM explicitly I think

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 7 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: Me neither xD maybe it's too german: Weder noch!

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 5 above)
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: you mean the emoji or the whole "(and maybe boring ...)"?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 7 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: I think of something else

Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: And since it claims to be both it is neither the one nor the other to a satisfying extend?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 9 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: Since Elixir is based on immutable data, you could not have started as [shown] in the first JavaScript solution.

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 5 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: the whole

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+title: "Programming beyond the comfort zone: JavaScript |> Elixir"
+date: "2017-02-02 07:00:00"
+image: blog-header/drink.jpg
+categories: tech-stories
+authors: ["Erik"]
+---
+
+I'm a JavaScript Engineer.
+I write JavaScript.
+I really like writing JavaScript and I feel very comfortable doing so.
```
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: This is just so everything is on a separate line for diffing right? Nothing to do with linebreaks does it?

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 3 above)
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: take either of the above

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md (hunk 11 above)
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: it has nothing to do with linebreaks;) just to keep the markdown clearer

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+I won't bother you anymore with advantages or language features Elixir provides nor try to persuade you that this is the right language to learn and use.
+Anyway, you should probably check some resources and dive a bit deeper into the awesome world of functional programming.
+A good place to start is [the official elixir-lang site](http://elixir-lang.org/).
+Just click around and look at some beautiful code and examples.
+
+You think Elixir is not for you?
+Check out another functional language then!
+Take a look at [Scala](https://www.scala-lang.org/), [Haskell](https://www.haskell.org/) or Erlang itself.
+I personally think your daily (programming) life will benefit from everything you learn, even if you do not actively use these languages for active development.
```
Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: two times "active".. could we maybe delete the second "active"?

Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: lol yes of course delete the second

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+# or without explicitely naming the parameters
+Enum.reduce([1, 2, 3, 5, 8], &(&1 + &2)) |> IO.puts
+```
+
+JS
+
+```js
+// Printing the sum of an array
+console.log([1, 2, 3, 5, 8].reduce((item, acc) => acc + item))
```
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: `console.log([1, 2, 3, 5, 8].reduce((acc, item) => acc + item))`

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+```js
+// We all do love fibonacci!
+function fibonacci (n) {
+   if (n < 2) return 1
+
+   return fibonacci(n - 2) + fibonacci(n - 1)
+}
+```
+
+###Enough of the platform
```
Comment (BirgitBader) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: ### Enough ... (insert blank space after the hashes)

**Comment by roseritter** [comment]
@BirgitBader i just finished up the pending adjustments that were to make. 
Do you want to do the final check?:)

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+I won't bother you anymore with advantages or language features Elixir provides nor try to persuade you that this is the right language to learn and use.
```
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: At this point I'll stop listing language features Elixir provides since this should only be some kind of sweetening. I will also not try to persuade you that [...]

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+function fibonacci (n) {
+   if (n < 2) return 1
+
+   return fibonacci(n - 2) + fibonacci(n - 1)
+}
+```
+
+### Enough of the platform
+
+I won't bother you anymore with advantages or language features Elixir provides nor try to persuade you that this is the right language to learn and use.
```
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: At this point I'll stop listing language features Elixir provides since this post is only meant for sweetening. I will also not try to persuade you that [...]

Comment (roseritter) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: thanks, Erik! I adjusted that:)

**Comment by erikmueller** [comment]
🚢

File: _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md [hunk 17]
```diff
@@ -19,7 +19,7 @@ So why should you start programming in another language?
 
 Let's take a step back from the frameworks and look at the language itself.
 JavaScript claims to be both, object-oriented (with prototypal inheritance) and functional.
-And since it claims to be both, it is neither the one nor the other to a satisfying extend.
+And since it claims to be both, it is neither one or the other to a satisfying extend.
```
Comment (erikmueller) [inline _posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md]: Just noticed while reading but why was that changed?
